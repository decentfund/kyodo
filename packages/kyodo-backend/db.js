const mongoose = require('mongoose');
const { SystemError } = require('./errors');

const Schema = mongoose.Schema;

const Period = require('./models/Period');
const User = require('./models/User');
const Tip = require('./models/Tip');

mongoose.connection.on('error', e => {
  if (e.message.code === 'ETIMEDOUT') {
    console.log(e);
  }
  console.log(e);
});

mongoose.connection.once('open', () => {
  console.log(`MongoDB successfully connected to mongoUri`);
});

const initDb = () => mongoose.connect('mongodb://localhost/colony');

const taskSchema = new mongoose.Schema({
  taskId: String,
  taskTitle: String,
  specificationHash: String,
  finalized: Boolean,
  cancelled: Boolean,
  dateCreated: Date,
  dueDate: Date,
  potId: Number,
  // domainId: Number,
  // id: Number,
  // skillId: Number
});

const colonySchema = new mongoose.Schema({
  colonyName: String,
  colonyId: String,
  colonyAddress: String,
  tokenAddress: String,
  tokenName: String,
  tokenSymbol: String,
  creationBlockNumber: Number,
  creationDate: Date,
  periodIds: Array,
});

colonySchema.virtual('currentPeriodId').get(function() {
  return this.periodIds[this.periodIds.length - 1];
});

const domainSchema = new Schema({
  domainId: Number,
  domainTitle: String,
  localSkillId: Number,
  potId: Number,
});

const Task = mongoose.model('Task', taskSchema);
const Colony = mongoose.model('Colony', colonySchema);
const Domain = mongoose.model('Domain', domainSchema);

const getColonyById = async colonyId => {
  return await Colony.findOne({ colonyId });
};

const getCurrentUserPeriod = async (alias, periodId) => {
  const user = await User.findOne({ alias });
  const period = await Period.findOne({ user: user._id, periodId });
  const sentTips = await Tip.find({ from: user, periodId });
  const usedPoints = sentTips.reduce((sum, v) => sum + v.amount, 0);

  if (!period) throw new SystemError('Cannot find user current period');

  return {
    ...period,
    initialBalance: period.balance,
    balance: period.balance - usedPoints,
  };
};

const getDomainByCode = async code => {
  const domain = await Domain.findOne({ domainTitle: code });
  if (!domain) throw Error('Domain not found');

  return domain;
};

module.exports = {
  initDb,
  Task,
  Colony,
  Domain,
  Tip,
  User,
  Period,
  getColonyById,
  getCurrentUserPeriod,
  getDomainByCode,
};
