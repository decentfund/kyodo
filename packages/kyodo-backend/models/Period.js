import mongoose, { Schema } from 'mongoose';

const periodSchema = new Schema({
  title: String,
  address: String,
  periodId: Number,
  balance: Number,
  initialBalance: Number,
  user: { type: Schema.Types.ObjectId, ref: 'User' },
});

const Period = mongoose.model('Period', periodSchema);
export default Period;
