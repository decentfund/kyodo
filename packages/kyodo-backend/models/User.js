import mongoose, { Schema } from 'mongoose';

const userSchema = new Schema({
  address: String,
  alias: String,
  aliasSet: Number,
  balance: Number,
  domains: Array,
  tasks: Array,
  dateCreated: Date,
});

const User = mongoose.model('User', userSchema);
export default User;
