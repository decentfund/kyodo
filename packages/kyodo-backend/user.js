import web3 from 'web3';
import { Colony, getColonyById, getCurrentUserPeriod } from './db';
import User from './models/User';
import { createAndSaveNewUserPeriod } from './period';
const Point = require('./models/Point');
const { ModelError } = require('./errors');

export const dbAddUser = async ({
  address,
  alias,
  balance,
  domains,
  tasks,
}) => {
  const user = new User({
    address,
    alias,
    balance,
    domains,
    tasks,
    dateCreated: Date.now(),
  });

  await user.save(err => {
    if (err) return null;
  });

  return user;
};

export const addUser = async ({ alias }) => {
  const colony = await Colony.findOne();
  const user = await dbAddUser({ alias });
  await createAndSaveNewUserPeriod({
    periodId: colony.currentPeriodId,
    balance: 0,
    user,
  });
  return user;
};

export const dbGetAllUsers = async () => {
  return await User.find((err, users) => {
    if (err) {
      console.log(err);
      return [];
    }
    return users;
  });
};

export const getAllUsers = async (req, res) => {
  let users = await dbGetAllUsers();
  res.status(200).send(users);
  return users;
};

export const getUserByAddress = async (req, res) => {
  await User.find({ address: req.body.address });
};

export const getUserByAlias = async (req, res) => {
  await User.find({ alias: req.body.alias });
};

export const dbGetUserByAlias = async alias => {
  const user = await User.findOne({ alias });

  if (!user) {
    throw new ModelError('Not found', 'user', alias);
  }
  return user;
};

export const getOrCreateUser = async ({ alias }) => {
  let user;
  try {
    user = await dbGetUserByAlias(alias);
  } catch (e) {
    if (e.message === 'Not found') {
      user = await addUser({ alias });
    }
  }
  return user;
};

export const delegatePoints = async ({ amount, sender, receiver }) => {
  // Verify sender is present and has enough points
  if (!sender) throw Error('No sender specified');
  // Throw error if sender is not present
  const senderUser = await dbGetUserByAlias(sender);

  const balance = await getUserBalance(sender);

  if (balance < amount)
    throw Error(
      `Naaah, you don't have so much points. Your current point balance is ${balance}`,
    );

  // Get receiver if not present create a new one
  const receiverUser = await getOrCreateUser({ alias: receiver });

  // Create points and pass to user
  const points = new Point({
    amount,
    used: false,
    delegatee: senderUser,
    owner: receiverUser,
  });

  await points.save();

  return points;
};

export const getUserBalance = async alias => {
  const colony = await getColonyById(0);
  const currentPeriodId = colony.periodIds[colony.periodIds.length - 1];
  const { balance } = await getCurrentUserPeriod(alias, currentPeriodId);
  return balance;
};

export const setUserAlias = async ({ user, alias, blockNumber }) => {
  user.alias = alias;
  user.aliasSet = blockNumber;

  user.save(err => {
    if (err) return null;
  });

  return user;
};

export const addDomainToUser = async (req, res) => {
  // TODO:
};

export const addTaskToUser = async (req, res) => {
  // TODO:
};

export const changeUserAlias = async (req, res) => {
  // TODO:
  // this.getUserByAlias(req.body.alias);
};

export const updateUserAddress = async ({ alias, address }) => {
  if (!web3.utils.isAddress(address))
    throw Error('Address is incorrect, try again');
  const user = await dbGetUserByAlias(alias);
  if (!user) throw Error('Get some tips to get started');
  if (user.address)
    throw Error('Address is already set, I cannot change it by now');

  user.address = address;
  await user.save();
  return user;
};
