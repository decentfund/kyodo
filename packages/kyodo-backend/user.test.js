import mongoose from 'mongoose';
import { Domain, Colony, getCurrentUserPeriod } from './db';
import * as db from './db';
import { addUser, delegatePoints } from './user';
const Point = require('./models/Point');
import User from './models/User';
import Period from './models/Period';

describe('insert', () => {
  let connection;

  beforeAll(async () => {
    connection = await mongoose.connect(global.__MONGO_URI__);

    const govDomain = new Domain({ domainTitle: 'GOV' });
    await govDomain.save();

    const colony = new Colony({ colonyId: 0, periodIds: [0] });
    await colony.save();

    // await User.getUpdate();
  });

  afterAll(async () => {
    await connection.disconnect();
  });

  afterEach(async () => {
    await User.remove({});
    await Period.remove({});
  });

  beforeEach(async () => {
    const user = await addUser({ alias: 'igor' });

    // Increasing user balance
    await User.update(
      { _id: user._id },
      { balance: 10, address: '0x0' },
      console.log,
    );

    const currentUserPeriod = await Period.findOne({
      user: user._id,
      periodId: 0,
    });

    await Period.update({ _id: currentUserPeriod._id }, { balance: 10 });
  });

  describe('send new tip', () => {
    const knownUser = {
      alias: 'igor',
      address: '0x0',
    };
    const unknownUser = {
      alias: 'alina',
    };
    const existingUserMock = new User(knownUser);
    const newUserMock = new User(unknownUser);
    const receiver = newUserMock;
    it('delegates points properly', async () => {
      await delegatePoints({
        sender: 'igor',
        amount: 5,
        receiver: receiver.alias,
      });

      const points = await Point.findOne();
      expect(points.amount).toEqual(5);
      const { balance: currentUserBalance } = await getCurrentUserPeriod(
        'igor',
        0,
      );
      expect(currentUserBalance).toBe(5);
    });
    it('delegates points properly several times', async () => {
      await delegatePoints({
        sender: 'igor',
        amount: 5,
        receiver: receiver.alias,
      });

      await delegatePoints({
        sender: 'igor',
        amount: 3,
        receiver: receiver.alias,
      });

      const { balance: currentUserBalance } = await getCurrentUserPeriod(
        'igor',
        0,
      );
      expect(currentUserBalance).toBe(2);
    });
    it('if user does not have enough balance', async () => {
      try {
        await delegatePoints({
          sender: 'igor',
          amount: 15,
          receiver: receiver.alias,
        });
      } catch (e) {
        expect(e.message).toMatch(
          "Naaah, you don't have so much points. Your current point balance is 10",
        );
      }
    });
    it('displays proper points on delegate', async () => {
      await delegatePoints({
        sender: 'igor',
        amount: 5,
        receiver: receiver.alias,
      });

      const { balance: currentUserBalance } = await getCurrentUserPeriod(
        receiver.alias,
        0,
      );
      expect(currentUserBalance).toBe(5);
    });
  });
});
