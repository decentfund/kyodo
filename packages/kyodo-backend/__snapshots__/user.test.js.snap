// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generates final activity tasks properly works correctly 1`] = `
Array [
  Object {
    "amount": 5,
    "domain": Object {
      "domainTitle": "GOV",
    },
    "from": Object {
      "address": "0x0",
      "alias": "igor",
      "balance": 10,
    },
    "to": Object {
      "address": "0x0",
      "alias": "pasha",
      "balance": 10,
    },
  },
  Object {
    "amount": 3,
    "domain": Object {
      "domainTitle": "GOV",
    },
    "from": Object {
      "address": "0x0",
      "alias": "igor",
      "balance": 10,
    },
    "to": Object {
      "address": "0x0",
      "alias": "pasha",
      "balance": 10,
    },
  },
  Object {
    "amount": 1,
    "domain": Object {
      "domainTitle": "GOV",
    },
    "from": Object {
      "address": "0x0",
      "alias": "igor",
      "balance": 10,
    },
    "to": Object {
      "address": "0x0",
      "alias": "pasha",
      "balance": 10,
    },
  },
  Object {
    "amount": 3,
    "domain": Object {
      "domainTitle": "GOV",
    },
    "from": Object {
      "address": "0x0",
      "alias": "pasha",
      "balance": 10,
    },
    "to": Object {
      "address": "0x0",
      "alias": "igor",
      "balance": 10,
    },
  },
]
`;
